build: late and confusing error when  docker or podman are not available

**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

On the current `master` branch, hen neither docker nor podman is available and working on the system, developer operations that depend on containers for exectution,  like `make build`, `make lint.yaml`, and `make lint.markdown` fail with confusing error messages like `build: command not found` or `run: command not found`, sometimes (too) late in the process. 


**Expected behavior:**

T he mentioned  commands (`make build`, `make lint.yaml`,  and `make lint.markdown` should fail early with a clear message that neither docker nor podman are available or working.



**How to reproduce it (minimal and precise):**

On a system without functioning podman or docker available (like macOS with podman machine or docker desktop not started), `cd` into a rook git clone directory, make sure the `master` branch is checked out and current,  run the  above-mentioned commands, and observe the output:


**prepare  working directory:**
```shell
$ cd $(go env GOPATH)/src/github.com/rook/rook
$ git checkout master
...
$ git remote -v | grep origin
origin	git@github.com:rook/rook.git (fetch)
origin	git@github.com:rook/rook.git (push)
$ git fetch origin
...
$ git reset --hard origin/master
```

**lint.yaml:**

```shell

$ make lint.yaml
run  --rm -t -v /Users/obnox/go/src/github.com/rook/rook:/workdir:ro -w /workdir --entrypoint yamllint cytopia/yamllint@sha256:3e9eb827ab2b12a5ea5f49d4257bb3aca94bba9f1ba427c8bc7f2456385a5204 -c .yamllint deploy/examples/ --no-warnings
bash: line 1: run: command not found
make: *** [lint.yaml] Error 127
```

**lint.markdown:**

```shell
$ make lint.markdown 
bash: line 1: run: command not found
make: *** [lint.markdown] Error 127
```

***build:***

```shell
$ make build
...
rm -f "/tmp/rook-ceph-image-list"
Go version: go version go1.26.0 darwin/arm64
=== go build linux_arm64
=== container build build-5d8ac50f/ceph-arm64
bash: line 1: build: command not found
make[2]: *** [do.build] Error 127
make[1]: *** [ceph.linux_arm64] Error 2
make: *** [build] Error 2
$ 
```

This issue requests to change the behavior to let `make` fail with a clear error message that a working container runtime tool (docker or podman) is not found (or provided with the `DOCKERCMD` variable). for targets that require running containers like `lint.taml`, `lint.markdown`, 
 and `build`.

Additionally, a less aggressive behavior is suggested for targets `build` and `clean` that have a container image step:

instead of letting `build`and `clean` fail hard at the container image step, let them skip the image step gracefully with a clear informational message.




 





**File(s) to submit**:

not applicable. See output pasted above.

**Logs to submit**:

* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

none applicable.

**Cluster Status to submit**:

not applicable.

* Output of kubectl commands, if necessary

not applicable

**Environment**:
* OS: macOS 26.5.2
* Kernel: irrelevant, but `Darwin obibmac.local 25.5.0 Darwin Kernel Version 25.5.0: Tue Jun  9 22:28:29 PDT 2026; root:xnu-12377.121.10~1/RELEASE_ARM64_T6030 arm64`
* * Rook version: current master branch
* Storage backend version: not applicable
* 
* Kubernetes cluster type : none
* Storage backend status: not applicable

--- a/Makefile
+++ b/Makefile
@@ -197,19 +197,19 @@ lint.go: lint.fmt lint.vet golangci-lint ## run various go linters
 
 
 .PHONY: lint.markdown
-lint.markdown: ## Check formatting of documentation sources
+lint.markdown: check.container.runtime ## Check formatting of documentation sources
 	@$(MARKDOWNLINT) "Documentation/**/**.md" "#Documentation/Helm-Charts/**" --config .markdownlint-cli2.cjs
 
 .PHONY: lint.markdown-links
-lint.markdown-links:
+lint.markdown-links: check.container.runtime
 	@$(DOCKERCMD) run --rm -v "$$PWD:/workspace" --entrypoint "/workspace/tests/scripts/check-markdown-links-internal.sh" ghcr.io/tcort/markdown-link-check:stable
 
 .PHONY: fix.markdown
-fix.markdown: ## Check and fix formatting of documentation sources
+fix.markdown: check.container.runtime ## Check and fix formatting of documentation sources
 	@$(MARKDOWNLINT) "Documentation/**/**.md" "#Documentation/Helm-Charts/**" --fix --config .markdownlint-cli2.cjs
 
 .PHONY: lint.yaml
-lint.yaml: ## lint yaml files
+lint.yaml: check.container.runtime ## lint yaml files
 	$(YAMLLINT) -c .yamllint deploy/examples/ --no-warnings
 
 .PHONY: lint.helm

--- a/build/makelib/common.mk
+++ b/build/makelib/common.mk
@@ -35,6 +35,10 @@ DOCKERCMD=$(shell podman version >/dev/null 2>&1 && echo podman)
 endif
 endif
 
+.PHONY: check.container.runtime
+check.container.runtime:
+	$(if $(shell $(DOCKERCMD) version >/dev/null 2>&1 && echo available),,$(error no usable container runtime found. Install docker or podman, or set DOCKERCMD to a working command.))
+
 MARKDOWNLINT := $(DOCKERCMD) run --rm -v $$PWD:/workdir davidanson/markdownlint-cli2:$(MARKDOWNLINT_IMAGE_VERSION)
 
 ifneq ($(strip $(YAMLLINT_IMAGE_SHA)),)

--- a/images/ceph/Makefile
+++ b/images/ceph/Makefile
@@ -60,14 +60,16 @@ do.build:
 	@mkdir -p $(BUILD_CONTEXT_DIR)/rook-external/test-data
 	@cp $(MANIFESTS_DIR)/create-external-cluster-resources.* $(BUILD_CONTEXT_DIR)/rook-external/
 	@cd $(BUILD_CONTEXT_DIR) && $(SED_IN_PLACE) 's|BASEIMAGE|$(BASEIMAGE)|g' Dockerfile
-	@if [ "$(BUILD_CONTAINER_IMAGE)" != "false" ]; then\
+	@if [ "$(BUILD_CONTAINER_IMAGE)" = "false" ]; then\
+		echo "=== skipping container build because BUILD_CONTAINER_IMAGE=false"; \
+	elif ! $(DOCKERCMD) version >/dev/null 2>&1; then \
+		echo "=== skipping container build because no usable container runtime was found"; \
+	else \
 		$(DOCKERCMD) build --platform linux/$(GOARCH) $(BUILD_ARGS) \
 		--build-arg S5CMD_VERSION=$(S5CMD_VERSION) \
 		--build-arg S5CMD_ARCH=$(S5CMD_ARCH) \
 		-t $(CEPH_IMAGE) \
 		$(BUILD_CONTEXT_DIR);\
-	else \
-		echo "=== skipping container build because BUILD_CONTAINER_IMAGE=false"; \
 	fi
 	@if [ -z "$(SAVE_BUILD_CONTEXT_DIR)" ]; then\
 		rm -fr $(BUILD_CONTEXT_DIR);\

--- a/images/image.mk
+++ b/images/image.mk
@@ -75,9 +75,16 @@ BUILD_BASE_ARGS += $(BUILD_ARGS)
 all: build
 
 build: do.build ## Build images for the host platform.
-	@$(MAKE) cache.images
+	@if [ "$(BUILD_CONTAINER_IMAGE)" != "false" ] && $(DOCKERCMD) version >/dev/null 2>&1; then \
+		$(MAKE) cache.images; \
+	fi
 
-clean: clean.build ## Remove all images created from the current build.
+clean: ## Remove all images created from the current build.
+	@if $(DOCKERCMD) version >/dev/null 2>&1; then \
+		$(MAKE) clean.build; \
+	else \
+		echo "=== skipping image cleanup because no usable container runtime was found"; \
+	fi
 
 prune: cache.prune ## Prunes orphaned and cached images at the host level.
 

--- a/pkg/operator/ceph/object/objectstore.go
+++ b/pkg/operator/ceph/object/objectstore.go
@@ -933,7 +933,7 @@ func adjustZoneDefaultPools(objContext *Context, zone map[string]interface{}, sp
 		// default pool is not presented in shared pool spec
 		return zone, nil
 	}
-	// add zone namespace to metadata pool to safely share accorss rgw instances or zones.
+	// add zone namespace to metadata pool to safely share across rgw instances or zones.
 	// in non-multisite case zone name equals to rgw instance name
 	defaultMetaPool = defaultMetaPool + ":" + name
 	for pool, nsSuffix := range zonePoolNSSuffix {

